### PR TITLE
Integrate emoji parsing;

### PR DESCRIPTION
**Pull request by thorbert**

### CODE REVIEW

**Comment by thorbert** [comment]
It might be better to separate the AFINN and the emoji lists and let build/index.js build the combined product.

It'd probably also be better to preserve the token order  instead of tacking emojis on the end.

**Comment by peterpme** [comment]
👋 Can I help move this along?

File: lib/index.js [hunk 1]
```diff
@@ -34,12 +34,14 @@ module.exports = function (phrase, inject, callback) {
     }
 
     // Storage objects
-    var tokens      = tokenize(phrase),
+    var tokens      = [],
```
Comment (thisandagain) [inline lib/index.js]: Why the change here?

File: package.json [hunk 2]
```diff
@@ -19,7 +19,8 @@
     "test": "make test"
   },
   "dependencies": {
-    "lodash.assign": "4.1.0"
+    "lodash.assign": "4.1.0",
+    "spliddit": "^2.1.1"
```
Comment (thisandagain) [inline package.json]: Please use absolute version numbers (remove the `^`)

**Comment by thisandagain** [comment]
@peterpme Sure. Here is what I think we need:

- [ ] Add peer reviewed validation dataset which includes emojis
- [ ] Check to ensure that changes to the tokenizer included in this PR don't adversely impact performance too much
- [ ] Clean-up a few nitpicks from the PR

File: build/index.js [hunk 3]
```diff
@@ -5,35 +5,68 @@
  * @author Andrew Sliwinski <andrew@diy.org>
  */
 
+/*globals Promise:true*/
```
Comment (thisandagain) [inline build/index.js]: I'd rather not introduce promises to this. This should be simple enough to be broken down into smaller functions to avoid callback hell.

File: lib/tokenize.js [hunk 4]
```diff
@@ -5,10 +5,24 @@
...
-module.exports = function (input) {
-    return input
-        .toLowerCase()
-        .replace(/[^a-z0-9á-úñäâàéèëêïîöôùüûœç\- ]+/g, '')
-        .replace('/ {2,}/',' ')
-        .split(' ');
+var spliddit = require('spliddit');
+
+module.exports = function(input) {
+    if (input === undefined) throw new Error('input undefined');
```
Comment (thisandagain) [inline lib/tokenize.js]: This module should never throw.

File: lib/tokenize.js [hunk 5]
```diff
@@ -5,10 +5,24 @@
...
+    if (input.trim() === '') return [];
+    var tokens = spliddit(input.toLowerCase()
```
Comment (thisandagain) [inline lib/tokenize.js]: This code path could use some documentation / line comments.

**Comment by thisandagain** [comment]
Resolves https://github.com/thisandagain/sentiment/issues/24

File: package.json [hunk 6]
```diff
@@ -19,7 +19,10 @@
   "scripts": {
     "test": "make test"
   },
-  "dependencies": {},
+  "dependencies": {
+    "async": "^2.1.5",
```
Comment (thisandagain) [inline package.json]: Because `async` is only used by the build script it should only be included in `devDependencies`.

**Comment by thisandagain** [comment]
I was able to find a listing of emoji sentiment that we may want to use as the source (similar to how we use AFINN) for the `sentiment` module. I think I'd be more comfortable landing support with this as the backing dataset.

### Overview
http://kt.ijs.si/data/Emoji_sentiment_ranking/index.html

### Dataset
https://www.clarin.si/repository/xmlui/handle/11356/1048

### Paper
http://journals.plos.org/plosone/article?id=10.1371/journal.pone.0144296

File: build/emoji.txt [hunk 7]
```diff
@@ -0,0 +1,79 @@
+😁	5
```
Comment (thisandagain) [inline build/emoji.txt]: We should drop this file now that we are including `Emoji_Sentiment_Data_v1.0.csv`

**Comment by thisandagain** [comment]
Thanks @thorbert ! This is really getting close. I think I just have two remaining comments and then we should land / release.

**Comment by thisandagain** [comment]
Thanks @thorbert ! Just pulling everything down for testing now. Two things I'm observing:

- Performance: the new tokenizer has reduced performance by almost 50%
- Test coverage: test coverage has dropped from 100% to 87.5% (also due to changes in the tokenizer)

I'll keep digging to see if this might be easy to resolve.

**Comment by thisandagain** [comment]
It appears that the use of `spliddit.hasPair` is the main performance culprit. Particularly where you are attempting to tokenize emoji that are not delimited by a space (e.g. `😀💩👻`). This also happens to be where you don't have test coverage.

I don't really see a way around this completely, but having better unicode support for the tokenizer might be worth the performance cost. Here is a quick first shot at attempting to optimize and improve readability using a flat map:
```js
var spliddit = require('spliddit');

/**
 * Applys a "flatten" reducer to the standard JS "map" array method.
 * @param  {Function} lambda Reducer
 * @return {array}           Flattened array
 */
Array.prototype.flatMap = function(lambda) {
    return Array.prototype.concat.apply([], this.map(lambda));
};

/**
 * Remove special characters and return an array of tokens (words).
 * @param  {string} input Input string
 * @return {array}        Array of tokens
 */
module.exports = function(input) {
    // Normalize input
    input = input
        .toLowerCase()
        .replace(/[.,\/#!$%\^&\*;:{}=_`\"~()]/g, '')
        .replace('/ {2,}/',' ');

    // Split on spaces
    input = spliddit(input, ' ');

    // Iterate over tokens and split unicode pairs
    return input.flatMap(function (t) {
        if (!spliddit.hasPair(t)) return t;
        return spliddit(t, '');
    });
};
```

Another major issue I noticed while testing is that something is amiss with the build script. Common emoji like 😀 and 💩 are included in `Emoji_Sentiment_Data_v1.0.csv` but are not present in the resulting JSON used by sentiment.

**Comment by thorbert** [comment]
Yes, benchmarks don't look good. I can expand tests but I'm not sure my toy use case makes sense generally (SMS messages from a 4 year old which contain only undelimited emoji). Maybe we can toss that out if doing so helps to restore the performance measures.

**Comment by thisandagain** [comment]
@thorbert Sure. Let's remove that change to the tokenizer. We could always restore it later as a configuration option, but I wouldn't want all users to take on that performance hit.

File: lib/tokenize.js [hunk 8]
```diff
@@ -5,10 +5,24 @@
...
-module.exports = function (input) {
-    return input
-        .toLowerCase()
-        .replace(/[^a-z0-9á-úñäâàéèëêïîöôùüûœç\-\' ]+/g, '')
-        .replace('/ {2,}/',' ')
-        .split(' ');
+var spliddit = require('spliddit');
+
+module.exports = function(input) {
+    if (input === undefined) return [];
```
Comment (thisandagain) [inline lib/tokenize.js]: Because we check if the input if `undefined` in the containing module (https://github.com/thisandagain/sentiment/blob/develop/lib/index.js#L42), this can be removed.

File: lib/tokenize.js [hunk 9]
```diff
@@ -5,10 +5,24 @@
...
+    if (input.trim() === '') return [''];
```
Comment (thisandagain) [inline lib/tokenize.js]: Is your thinking that this is a performance optimization? If so, I think I'd rather track it in a separate PR.

File: test/integration/sync_positive_emoji.js [hunk 10]
```diff
@@ -0,0 +1,14 @@
+var test = require('tap').test;
+var sentiment = require('../../lib/index');
+
+var dataset = 'This is so cool ✓';
```
Comment (thisandagain) [inline test/integration/sync_positive_emoji.js]: It would be nice to also include a test case with some more common emoji such as 😄 (that is how I found the build issue).

**Comment by thisandagain** [comment]
This is great! Thanks @thorbert for all your patience as we worked through the details. I'll land this now and then prep a release. 😄 👍
